Type pattern in switch expression involving generic type with non-keyword type parameters causes syntax error

**Version Used**: Compiler version: '3.8.0-4.20460.4 (64f2392e)'. Language version: preview.

**Steps to Reproduce**:

Compile this code: ([SharpLab](https://sharplab.io/#v2:EYLgtghgzgLgpgJwDQxAN0QSwGYE8A+AAgEwCMAsAFBWEDMABCfQCpyxUDeV9PjDhpAAz0AsgAoA9sABWcAMYx6EgJTde6gLwA+JfSgB3TDDkALNep5dKFm4yH1t9AEQBlGAkwA7AOZOk52x4BYgcdJ2C/AMDggB4vGB1HJwBBOSMJTzjPBMjrQN5YqVkFRLDCmXkc/zz8u2IYgEls0gA2UudYpphWrVzaoNJ61lh28MGY4aqo2wB9UOcAVU8Aa08JfU8naZ4AXwBuKh2qIA))

```csharp
using System;

public class Test
{
    public string M(object o)
        => o switch
        {
            string => "String",
            Action => "Action",
            Action<int> => "Aciton<int>",
            Action<object> => "Action<object>",
            Action<Int16> => "Action<Int16>", // <-- Syntax Errors
            Action<Test> => "Action<Test>", // <-- Syntax Errors
            _ => "Unknown"
        };
}
```

**Expected Behavior**: Compiles without errors

**Actual Behavior**: The lines marked have errors associated with them:

```
Line 12: error CS1001: Identifier expected
Line 12: error CS1003: Syntax error, '=>' expected
Line 12: error CS1525: Invalid expression term ','
Line 13: error CS1001: Identifier expected
Line 13: error CS1003: Syntax error, '=>' expected
Line 13: error CS1525: Invalid expression term ','
```

-------------

This does not happen with the equivalent switch statement: ([SharpLab](https://sharplab.io/#v2:EYLgtghgzgLgpgJwDQxAN0QSwGYE8A+AAgEwCMAsAFBWEDMABCfQCpyxUDeV9PjDhpAAz0AsgAoA9sABWcAMYx6EgJTdeXSry30oAd0ww5AC3qTVm7Tw2XLc6HEZCQamzcIB2egCIAyjASYAHYA5l4A3C6u9HZQDgLEzhZR2h7e8eGRrjFxpMQAPEEwAHyJyZapXgCCcgYSgQWBxRlJUdmO+VKyCiWZZRXVtfWd8k0RLVn27XkAko2kAGw9432eVTUwdTNzi81l0ZPxeaywS3ta/eubx6O9lgAmcNgQAK4ANqi3URUAqoEA1oEJLpArsbABfSIQyhQoA))

```csharp
using System;

public class Test
{
    public string M(object o)
    {
        switch (o)
        {
            case string:
                return "String";
            case Action:
                return "Action";
            case Action<int>:
                return "Aciton<int>";
            case Action<object>:
                return "Aciton<object>";
            case Action<Int16>:
                return "Aciton<Int16>";
            case Action<Test>:
                return "Aciton<Test>";
            default:
                return "Unknown";
        }
    }
}
```

--- a/src/Compilers/CSharp/Portable/Parser/LanguageParser.cs
+++ b/src/Compilers/CSharp/Portable/Parser/LanguageParser.cs
@@ -5559,6 +5559,10 @@ tryAgain:
                     // This is useful for parsing expressions in isolation
                     return ScanTypeArgumentListKind.DefiniteTypeArgumentList;
 
+                case SyntaxKind.EqualsGreaterThanToken:  // e.g. `e switch { A<B> => 1 }`
+                    // This token was added to 7.5.4.2 Grammar Ambiguities in C# 9.0
+                    return ScanTypeArgumentListKind.DefiniteTypeArgumentList;
+
                 default:
                     return ScanTypeArgumentListKind.PossibleTypeArgumentList;
             }

--- a/src/Compilers/CSharp/Test/Syntax/Parsing/PatternParsingTests.cs
+++ b/src/Compilers/CSharp/Test/Syntax/Parsing/PatternParsingTests.cs
@@ -11211,5 +11211,86 @@ switch (e)
                 EOF();
             }
         }
+
+        [Fact, WorkItem(47614, "https://github.com/dotnet/roslyn/issues/47614")]
+        public void GenericTypeAsTypePatternInSwitchExpression()
+        {
+            UsingStatement(@"_ = e switch { List<Foo> => 1, List<Bar> => 2, };",
+                TestOptions.RegularWithPatternCombinators
+                );
+            N(SyntaxKind.ExpressionStatement);
+            {
+                N(SyntaxKind.SimpleAssignmentExpression);
+                {
+                    N(SyntaxKind.IdentifierName);
+                    {
+                        N(SyntaxKind.IdentifierToken, "_");
+                    }
+                    N(SyntaxKind.EqualsToken);
+                    N(SyntaxKind.SwitchExpression);
+                    {
+                        N(SyntaxKind.IdentifierName);
+                        {
+                            N(SyntaxKind.IdentifierToken, "e");
+                        }
+                        N(SyntaxKind.SwitchKeyword);
+                        N(SyntaxKind.OpenBraceToken);
+                        N(SyntaxKind.SwitchExpressionArm);
+                        {
+                            N(SyntaxKind.TypePattern);
+                            {
+                                N(SyntaxKind.GenericName);
+                                {
+                                    N(SyntaxKind.IdentifierToken, "List");
+                                    N(SyntaxKind.TypeArgumentList);
+                                    {
+                                        N(SyntaxKind.LessThanToken);
+                                        N(SyntaxKind.IdentifierName);
+                                        {
+                                            N(SyntaxKind.IdentifierToken, "Foo");
+                                        }
+                                        N(SyntaxKind.GreaterThanToken);
+                                    }
+                                }
+                            }
+                            N(SyntaxKind.EqualsGreaterThanToken);
+                            N(SyntaxKind.NumericLiteralExpression);
+                            {
+                                N(SyntaxKind.NumericLiteralToken, "1");
+                            }
+                        }
+                        N(SyntaxKind.CommaToken);
+                        N(SyntaxKind.SwitchExpressionArm);
+                        {
+                            N(SyntaxKind.TypePattern);
+                            {
+                                N(SyntaxKind.GenericName);
+                                {
+                                    N(SyntaxKind.IdentifierToken, "List");
+                                    N(SyntaxKind.TypeArgumentList);
+                                    {
+                                        N(SyntaxKind.LessThanToken);
+                                        N(SyntaxKind.IdentifierName);
+                                        {
+                                            N(SyntaxKind.IdentifierToken, "Bar");
+                                        }
+                                        N(SyntaxKind.GreaterThanToken);
+                                    }
+                                }
+                            }
+                            N(SyntaxKind.EqualsGreaterThanToken);
+                            N(SyntaxKind.NumericLiteralExpression);
+                            {
+                                N(SyntaxKind.NumericLiteralToken, "2");
+                            }
+                        }
+                        N(SyntaxKind.CommaToken);
+                        N(SyntaxKind.CloseBraceToken);
+                    }
+                }
+                N(SyntaxKind.SemicolonToken);
+            }
+            EOF();
+        }
     }
 }

--- a/src/Compilers/CSharp/Test/Syntax/Parsing/PatternParsingTests.cs
+++ b/src/Compilers/CSharp/Test/Syntax/Parsing/PatternParsingTests.cs
@@ -11215,7 +11215,7 @@ switch (e)
         [Fact, WorkItem(47614, "https://github.com/dotnet/roslyn/issues/47614")]
         public void GenericTypeAsTypePatternInSwitchExpression()
         {
-            UsingStatement(@"_ = e switch { List<Foo> => 1, List<Bar> => 2, };",
+            UsingStatement(@"_ = e switch { List<X> => 1, List<Y> => 2, };",
                 TestOptions.RegularWithPatternCombinators
                 );
             N(SyntaxKind.ExpressionStatement);
@@ -11247,7 +11247,7 @@ switch (e)
                                         N(SyntaxKind.LessThanToken);
                                         N(SyntaxKind.IdentifierName);
                                         {
-                                            N(SyntaxKind.IdentifierToken, "Foo");
+                                            N(SyntaxKind.IdentifierToken, "X");
                                         }
                                         N(SyntaxKind.GreaterThanToken);
                                     }
@@ -11272,7 +11272,7 @@ switch (e)
                                         N(SyntaxKind.LessThanToken);
                                         N(SyntaxKind.IdentifierName);
                                         {
-                                            N(SyntaxKind.IdentifierToken, "Bar");
+                                            N(SyntaxKind.IdentifierToken, "Y");
                                         }
                                         N(SyntaxKind.GreaterThanToken);
                                     }

